### PR TITLE
perf(semantic): reduce index_vec allocation

### PR DESCRIPTION
**Pull request by Dunqing**
Combining some SymbolTable fields to the `Symbol` struct, we can reduce allocating `IndexVec`. After my changes, I suddenly realized the api is much easier to use now. We no longer need the `symbol_id` to get the various field value you want, you just need the `Symbol`

### CODE REVIEW

**Comment by Dunqing** [comment]
* **#4266** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4266?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by overlookmotel** [comment]
Interesting that this change from struct-of-arrays to array-of-structs is a gain on benchmarks. However, there is another possible approach to tackling this which may be even better: https://github.com/oxc-project/backlog/issues/11

**Comment by Dunqing** [comment]
> Interesting that this change from struct-of-arrays to array-of-structs is a gain on benchmarks. However, there is another possible approach to tackling this which may be even better: [oxc-project/backlog#11](https://github.com/oxc-project/backlog/issues/11)
> 
> However, since this PR is a gain at present, we could merge it now and attempt [oxc-project/backlog#11](https://github.com/oxc-project/backlog/issues/11) later on?

This PR is just an attempt at my guess, I'm not sure we want to merge it. I know you will be optimizing the whole `semantic`, and you probably have further improvements that can be done better. So it's up to you whether to merge it or not!

**Comment by Boshen** [comment]
SoA improves write performance but regresses read performance (notice drop in linter).

I think we should keep SoA unless there is a win across the table.

**Comment by Dunqing** [comment]
> SoA improves write performance but regresses read performance (notice drop in linter).

There are further optimizations we can make in linter, Such as we can pass `Symbol` to `run_on_symbol`. This may give a small performance boost

File: crates/oxc_semantic/src/symbol.rs [hunk 1]
```diff
@@ -23,82 +23,130 @@ const TS_APPEND_CONTENT: &'static str = r#"
...
+#[derive(Debug)]
+#[cfg_attr(feature = "serialize", derive(Serialize, Tsify), serde(rename_all = "camelCase"))]
+pub struct Symbol {
+    pub span: Span,
+    pub name: CompactStr,
+    pub flag: SymbolFlags,
+    pub scope_id: ScopeId,
+    /// Pointer to the AST Node where this symbol is declared
+    pub declaration: AstNodeId,
+    pub resolved_references: Vec<ReferenceId>,
```
Comment (DonIsaac) [inline crates/oxc_semantic/src/symbol.rs:35]: Should we encapsulate these fields for future flexibility?

File: crates/oxc_semantic/src/symbol.rs [hunk 2]
```diff
@@ -23,82 +23,130 @@ const TS_APPEND_CONTENT: &'static str = r#"
 export type IndexVec<I, T> = Array<T>;
 "#;
 
+#[derive(Debug)]
+#[cfg_attr(feature = "serialize", derive(Serialize, Tsify), serde(rename_all = "camelCase"))]
+pub struct Symbol {
```
Comment (DonIsaac) [inline crates/oxc_semantic/src/symbol.rs:28]: Consider adding a `GetSpan` impl.

**Comment by Dunqing** [comment]
Closing this because we tried another better way to improve more in #4328

**Pull request closed without merging**